flowmachine/flowmachine/core/random.py

```diff
@@ -3,293 +3,475 @@
 from .table import Table
 
 
-class RandomBase:
+class _RandomGetter:
     """
-    Base class containing method to construct a query used to obtain a random sample from a table.
+    Helper class for pickling/unpickling of dynamic random classes.
+    (see https://stackoverflow.com/questions/1947904/how-can-i-pickle-a-nested-class-in-python/11493777#11493777)
     """
 
-    def _inheritance_check(self):
+    def __call__(self, query, sampling_method, params):
+        return query.random_sample(sampling_method, **params)
+
+
+class RandomBase(metaclass=ABCMeta):
+    """
+    Base class for queries used to obtain a random sample from a table.
+    """
+
+    def __init__(self, query, *, size=None, fraction=None, estimate_count=True):
+        if size is None and fraction is None:
+            raise ValueError(
+                f"{self.__class__.__name__}() missing 1 required argument: 'size' or 'fraction'"
+            )
+        if size is not None and fraction is not None:
+            raise ValueError(
+                f"{self.__class__.__name__}() expects only 1 argument to be defined: either 'size' or 'fraction'"
+            )
+
+        if fraction is not None and (fraction < 0 or fraction > 1):
+            raise ValueError(
+                f"{self.__class__.__name__}() expects fraction between 0 and 1."
+            )
+
+        self.query = query
+        self.size = size
+        self.fraction = fraction
+        self.estimate_count = estimate_count
+
+    @property
+    def _sample_params(self):
+        """
+        Parameters passed when initialising this query.
         """
-        Raise a value error if the query is a table, and has children, and the method relies
-         on it not having children.
+        return {
+            "size": self.size,
+            "fraction": self.fraction,
+            "estimate_count": self.estimate_count,
+        }
+
+    def _count_table_rows(self):
+        """
+        Return a count of the number of rows in self.query table, either using
+        information contained in the `pg_class` (if self.estimate_rowcount) or
+        by performing an actual count in the number of rows.
         """
-        if (self.method != "system_rows") or (not isinstance(self.query, Table)):
-            return
-        if self.query.has_children():
+        rowcount = 0
+
+        if self.estimate_count:
+            table = self.query.get_table()
+            rowcount = table.estimated_rowcount()
+
+        if not self.estimate_count or rowcount == 0:
+            rowcount = len(self.query)
+
+        return rowcount
+
+    @abstractmethod
+    def _make_query(self):
+        raise NotImplementedError(
+            f"Class {self.__class__.__name__} does not implement the _make_query method."
+        )
+
+    # Overwrite the table_name method so that it cannot
+    # be stored by accident.
+    @property
+    def table_name(self):
+        if hasattr(self, "seed") and self.seed is not None:
```

Comment:
Wondering if, rather than needing to check the logic in a bunch of places, it would be nicer to have a `SeededRandom` class to mix in, or `SeededRandom(RandomBase, metaclass=ABCMeta)` to use as another step in inheritance chain?